### DISCUSSION TITLE
Tutorial code raises mypy dict entry incompatible type errors

### DISCUSSION THREAD

**Discussion by yiyoascen** [fastapi/fastapi#8862] [Questions] [answered]
Didn't find a similar error issue, I'm using PopOs with mypy 0.812

from typing import Optional

from fastapi import FastAPI

app = FastAPI()

@app.get("/items/")

async def read_items(q: Optional[str] = None):

    results = {"items": [{"item_id": "Foo"}, {"item_id": "Bar"}]}
    if q:
        results.update({"q": q})
    return results

throws the following mypy error:
error: Dict entry 0 has incompatible type "str": "str"; expected "str": "List[Dict[str, str]]"

also this example raises a similar error: 
"""requestbody+ path + query params"""
from typing import Optional
from fastapi import FastAPI
from pydantic import BaseModel

class Item(BaseModel):
    name: str
    description: Optional[str] = None
    price: float
    tax: Optional[float] = None

app = FastAPI()

@app.put("/items/{item_id}")
async def create_item(item_id: int, item: Item, q: Optional[str] = None):
    result = {"item_id": item_id, **item.dict()}
    if q:

        result.update({"q": q})
    return result

error: Dict entry 0 has incompatible type "str": "str"; expected "str": "int"
Found 1 error in 1 file (checked 1 source file)

**Comment by ycd** [comment]
Could you please format the code?

**Comment by tiangolo** [comment, marked as answer]
Hello! If you still have problems, please create a new question following the template and filling in the details, it will help us understand the question and help you better. ☕

**Comment by qbicsky** [comment]
Hi!
Sorry for waking up such an old topic. But I've encountered the same issue.
This error message is related to enabled plugin called Mypy Type Checker (or any other related to Mypy linting).
Disabling Mypy will hide this error. Code works fine without it.

My question is why Mypy shows such an issue?

My code for that case is is:
```
@app.get("/items/")
async def read_items(q: Annotated[str | None, Query(max_length=50)] = None):
    results = {"items": [{"item_id": "Foo"}, {"item_id": "Bar"}]}
    if q:
        results.update({"q": q})
    return results
```

And a mypy message:
`Dict entry 0 has incompatible type "str": "str"; expected "str": "list[dict[str, str]]"`

**Reply by YuriiMotov** [reply]
Since you don't specify type for `results` variable, mypy tries to determine it from the expression (`{"items": [{"item_id": "Foo"}, {"item_id": "Bar"}]}`) and it thiks that it's `dict[str: list[dict[str, str]]]`.
Then you are trying to add str value to this dict instead of `list[dict[str, str]]`.

To solve this issue you can specify the type of `results` explicitly:
`results: dict[str, Any] = {"items": [{"item_id": "Foo"}, {"item_id": "Bar"}]}`

or
`results: dict[str, str | list[dict[str, str]]] = {"items": [{"item_id": "Foo"}, {"item_id": "Bar"}]}`

**Reply by qbicsky** [reply]
Whoa! That was a fast and precise response. :)
Gone with 1st approach. Problem solved. Thanks!
